tools/preprocess_pragmas.c

```diff
@@ -0,0 +1,146 @@
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+const char str_pragma_increment_block_number[] = "#pragma increment_block_number";
+
+int main(int argc, char** argv) {
+#if 0
+    // for debugging with a specific file
+#undef stdin
+    FILE* stdin = fopen("src/code/fault.c", "r");
+    assert(stdin != NULL);
+#endif
+
+    if (argc != 3) {
+        fprintf(stderr, "Usage: preprocess_pragmas OOT_VERSION filename < source.c\n");
+        return EXIT_FAILURE;
+    }
+    char* const version = argv[1];
+    const int len_version = strlen(version);
+    char* const filename = argv[2];
+
+    char buf[32 * 1024];
+    char* const bufend = buf + sizeof(buf);
+    char* bufp = buf;
+    size_t nread;
+    int line_num = 1;
+    // whether the current line follows a #pragma increment_block_number,
+    // including continuation lines (lines after a \-ending line)
+    bool is_in_pragma = false;
+    // the line where the #pragma increment_block_number is
+    int pragma_line_number;
+    // how many fake structs to write to replace the current pragma
+    int n_fake_structs;
+
+    while ((nread = fread(bufp, 1, bufend - bufp, stdin)) != 0) {
+        bufp += nread;
+        char* last_line_return = NULL;
+        for (char* p = bufp - 1; p >= buf; p--) {
+            if (*p == '\n') {
+                last_line_return = p;
+                break;
+            }
+        }
+        if (last_line_return == NULL) {
+            // No line return, read more data.
+            // Assert there is space for it (there should be no line long enough to not fit in buf).
+            assert(bufp < bufend);
+            continue;
+        }
+        char* line = buf;
+        while (true) {
+            char* line_end = line;
+            while (*line_end != '\n') {
+                line_end++;
+                assert(line_end <= last_line_return);
+            }
+            if (!strncmp(line, str_pragma_increment_block_number, strlen(str_pragma_increment_block_number))) {
+                is_in_pragma = true;
+                pragma_line_number = line_num;
+                n_fake_structs = 0;
+            }
+            if (is_in_pragma) {
+                *line_end = '\0';
+                char* version_amount_item = strstr(line, version);
+                if (version_amount_item != NULL) {
+                    if (version_amount_item[len_version] != ':') {
+                        fprintf(stderr, "Found version %s in pragma line but no :amount attached\n", version);
+                        fprintf(stderr, "%s\n", line);
+                        return EXIT_FAILURE;
+                    }
+                    char* version_amount_str_start = &version_amount_item[len_version + 1];
+                    char* version_amount_str_end;
+                    long amount = strtol(version_amount_str_start, &version_amount_str_end, 10);
```

Comment:
I think `strtol` will happily read garbage past the end of the buffer here? Say if the input line ends with `version:` since strtol discards leading whitespace. I'm not sure what to do about it, maybe just checking that the next char after `:` is a digit is enough

Reply:
the line has been NUL-terminated earlier with `*line_end = '\0';` so there should be no buffer overrun

I forgot strtol discards leading whitespace. I hate C text processing. Is this that big of a deal?